File: repos/app/wrapper/lcontest/LongContestExposedWrapper.R
library(fork)

currentTime <- as.integer(0)
startTime <- as.integer(0)

exposed_fd <- as.integer(13)
SIGXFSZ <- 25

reset <- function() {
    currentTime <- as.integer(0)
    startTime <- as.integer(0)
}

start <- function() {
    startTime <- gettime()
}

stop <- function() {
    end <- gettime()
    currentTime <- currentTime + (end - startTime)
}

getExposedTime <- function() {
    return (currentTime)
}

<EXPOSED_METHODS>
<METHOD_NAME> <- function(<PARAMS>):
    start()
    pid <- getpid()

    kill(pid, signal = SIGXFSZ)

    startMethod(exposed_fd, as.integer(<METHOD_NUMBER>))
    
    <WRITE_ARGS>
        <ARG_METHOD>(exposed_fd, <ARG_NAME>)
    </WRITE_ARGS>
        flush2(exposed_fd)
    
    val = <RETURN_METHOD_NAME>(exposed_fd)
    
    kill(pid, signal = SIGXFSZ)
    stop()
    
    return val
</EXPOSED_METHODS>

File: repos/app/wrapper/lcontest/LongContestExposedWrapper.py
import os
import sys
import time
import <WRAPPER_CLASS>

class Stopwatch:
	def __init__(self):
		self.time = 0
		self.startTime = 0
			
	def reset(self):
		self.time = 0
		self.startTime = 0
	
	def start(self):
		self.startTime = <WRAPPER_CLASS>.gettime()
	
	def stop(self):
		end = <WRAPPER_CLASS>.gettime()
		self.time = self.time + (end - self.startTime)
		self.startTime = 0
		
	def getTime(self):
		return self.time

exposed_fd = 13
SIGXFSZ = 25

watch = Stopwatch()

<EXPOSED_METHODS>
def <METHOD_NAME>(<PARAMS>):
	watch.start()
	
	os.kill(os.getpid(), SIGXFSZ)

	<WRAPPER_CLASS>.startMethod(exposed_fd, <METHOD_NUMBER>)
	
	<WRITE_ARGS>
	<WRAPPER_CLASS>.<ARG_METHOD>(exposed_fd, <ARG_NAME>)
	</WRITE_ARGS>
	<WRAPPER_CLASS>.flush(exposed_fd)
	
	val = <WRAPPER_CLASS>.<RETURN_METHOD_NAME>(exposed_fd)
	
	os.kill(os.getpid(), SIGXFSZ)
	watch.stop()
	
	return val
</EXPOSED_METHODS>
